compiler/hie-core/exe/Main.hs

```diff
@@ -128,10 +127,16 @@ showEvent lock (EventFileDiagnostics (toNormalizedFilePath -> file) diags) =
     withLock lock $ T.putStrLn $ showDiagnosticsColored $ map (file,) diags
 showEvent lock e = withLock lock $ print e
 
-newSession' :: Cradle -> IO HscEnv
+newSession' :: Cradle -> IO HscEnvEq
 newSession' cradle = getLibdir >>= \libdir -> do
     env <- runGhc (Just libdir) $ do
         initializeFlagsWithCradle "" cradle
         getSession
     initDynLinker env
-    pure env
+    newHscEnvEq env
+
+loadEnvironment :: FilePath -> IO (FilePath -> Action HscEnvEq)
+loadEnvironment dir = do
+    op <- newCacheIO $ \() -> do
+        liftIO $ newSession' =<< findCradle (dir <> "/")
+    return $ \_ -> op ()
```

Comment:
I assume this is just a place holder that will change with the new `hie-bios` API (which seems to be merged now)? If not, I don’t understand this code. You never actually use the file path. I am also not quite sure how the caching will work even with the the `hie-bios` changes. Do you just cache per file? That seems rather expensive given that this will mean that you will shell out to `cabal` (or whatever your project is) once per file which on a large project could take a long time.

Reply:
This is just a placeholder - the new hie-bios changes will enable its functionality. The idea using the new hie-bios API is that you call `findCradle` on every file (which returns the `hie.yaml`), caching at the level of `hie.yaml` you call `loadCradle`, and then you invoke the cradle on each file. The `loadCradle` is the thing that runs `cabal` so it should only run once per set of files under the same `hie.yaml` - so once if you just have one project.